login/pam.c

```diff
@@ -0,0 +1,272 @@
+    return failure(EXITCODE_PANIC, error_tag, "pam-prompt");
+  }
+
+  free(url);
+  url = NULL;
+
+  int bytes_read = strlen(input);
+  if (config->options & INSECURE) {
+    pam_syslog(pamh, LOG_DEBUG, "user input: %s", input);
+  }
+
+  // Calculate the correct authcode.
+  char authcode_encoded[ENCODED_BUFSIZE(sizeof authcode)] = {0};
+  if (base64url_encode(authcode, sizeof authcode, (uint8_t *)authcode_encoded,
+                       sizeof authcode_encoded) == 0) {
+    free(input);
+    return failure(EXITCODE_PANIC, error_tag, "authcode-encode");
+  }
+  if (config->options & INSECURE) {
+    pam_syslog(pamh, LOG_DEBUG, "expect input: %s", authcode_encoded);
+  }
+
+  if (bytes_read < MIN_ENCODED_AUTHCODE_LEN) {
+    free(input);
+    return failure(EXITCODE_INVALID_INPUT_SIZE, error_tag, "authcode-length");
+  }
+  if (bytes_read > strlen(authcode_encoded)) {
+    free(input);
+    return failure(EXITCODE_INVALID_INPUT_SIZE, error_tag, "authcode-length");
+  }
+
+  // Since we use (relatively) short auth codes, sleep before confirming the
+  // result to prevent bruteforcing.
+  struct timespec delay;
+  delay.tv_sec = (time_t)config->auth_delay_sec;
+  delay.tv_nsec = 0;
+  if (nanosleep(&delay, NULL) != 0) {
+    free(input);
+    return failure(EXITCODE_INTERRUPTED, error_tag, "auth-delay");
+  }
+
+  // We can use insecure strncmp() because it's a *one-time* code.
+  if (strncmp(input, authcode_encoded, bytes_read) != 0) {
+    free(input);
+    return failure(EXITCODE_INVALID_AUTHCODE, error_tag, "authcode-invalid");
+  }
+
+  free(input);
+  return 0;
+}
+
+int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc,
+                        const char **argv) {
+  const char *error_tag = NULL;
+  login_config_t config = {0};
+  int rc = PAM_AUTH_ERR;
+
+  int r = parse_pam_args(pamh, argc, argv, &config);
+  if (r < 0) {
+    pam_syslog(pamh, LOG_ERR, "failed to parse pam module arguments (%d)", r);
+    return rc;
+  }
+
+  r = parse_config_file(&config);
+  if (r < 0) {
+    pam_syslog(pamh, LOG_ERR, "failed to read config file: %s (%d)",
+               config.config_path, r);
+    return rc;
+  }
+
+  r = get_username(pamh, &config, &error_tag);
+  if (r < 0) {
+    pam_syslog(pamh, LOG_ERR, "failed to get username: %s (%d)", error_tag, r);
+    return rc;
+  }
+
+  r = postprocess_config(&config);
+  if (r < 0) {
+    pam_syslog(pamh, LOG_ERR, "failed to postprocess config (%d)", r);
+    return rc;
+  }
+
+  r = glome_authenticate(pamh, &config, &error_tag, argc, argv);
+  if (!r) {
+    rc = PAM_SUCCESS;
+    if (config.options & VERBOSE) {
+      pam_syslog(pamh, LOG_ERR, "authenticated user '%s'", config.username);
+    }
+  } else {
+    pam_syslog(pamh, LOG_ERR, "failed to authenticate user '%s': %s (%d)",
+               config.username, error_tag, r);
+  }
+
+  return rc;
+}
+
+int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char **argv) {
+  return PAM_SUCCESS;
```

Comment:
Added a comment for `PAM_SUCCESS` in the new version of the patch.

Reply:
But if we return `PAM_SUCCESS` here, applications will see `pam_setcred(3)` return with `PAM_SUCCESS` as well, won't they? According to the manual, this return value should be interpreted as _Data was successful stored_, and since we're not storing anything may be disingenuous.

Reply:
`PAM_CRED_UNAVAIL: This module cannot retrieve the user's credentials.` sounds correct to me.